File: stitcher_proxy/storage.py
import os
import json
import logging
from pathlib import Path
from typing import List, Dict, Any, Tuple
from .config import get_config

logger = logging.getLogger("stitcher.storage")

def get_session_dir(session_id: str) -> Path:
    config = get_config()
    session_dir = config.data_dir / session_id
    session_dir.mkdir(parents=True, exist_ok=True)
    return session_dir

def get_active_file(session_dir: Path) -> Path:
    return session_dir / "active.jsonl"

def append_messages(session_id: str, messages: List[Dict[str, Any]]):
    """Append messages to the active session file, rolling if necessary."""
    if not messages:
        return
        
    config = get_config()
    session_dir = get_session_dir(session_id)
    active_file = get_active_file(session_dir)
    
    # Check if we need to roll the file
    if active_file.exists() and active_file.stat().st_size >= config.roll_size_bytes:
        _roll_file(session_dir, active_file)
        
    with open(active_file, "a", encoding="utf-8") as f:
        for msg in messages:
            # We wrap the message in a record wrapper if needed, or just dump the message.
            # stitcher.py uses: {"type": "message", "message": msg}
            record = {"type": "message", "message": msg}
            f.write(json.dumps(record, ensure_ascii=False) + "\n")

def _roll_file(session_dir: Path, active_file: Path):
    """Rename active.jsonl to active.001.jsonl, etc."""
    existing_rolls = []
    for f in session_dir.iterdir():
        if f.name.startswith("active.") and f.name.endswith(".jsonl") and f.name != "active.jsonl":
            try:
                num = int(f.name.split(".")[1])
                existing_rolls.append(num)
            except ValueError:
                pass
                
    next_num = max(existing_rolls + [0]) + 1
    rolled_name = f"active.{next_num:03d}.jsonl"
    active_file.rename(session_dir / rolled_name)
    logger.info(f"Rolled session file to {rolled_name}")

def get_session_files(session_id: str) -> List[Path]:
    """Return all session files ordered from oldest to newest."""
    session_dir = get_session_dir(session_id)
    active_file = get_active_file(session_dir)
    
    rolled_files = []
    if session_dir.exists():
        for f in session_dir.iterdir():
            if f.name.startswith("active.") and f.name.endswith(".jsonl") and f.name != "active.jsonl":
                try:
                    num = int(f.name.split(".")[1])
                    rolled_files.append((num, f))
                except ValueError:
                    pass
                    
    rolled_files.sort()
    all_files = [path for _, path in rolled_files]
    if active_file.exists():
        all_files.append(active_file)
        
    return all_files

def get_stats() -> Dict[str, Any]:
    """Return global statistics about the storage."""
    config = get_config()
    session_count = 0
    total_messages = 0
    
    if config.data_dir.exists():
        for session_dir in config.data_dir.iterdir():
            if session_dir.is_dir():
                session_count += 1
                # Quick count of lines
                for f in session_dir.iterdir():
                    if f.name.endswith(".jsonl"):
                        try:
                            with open(f, "r", encoding="utf-8") as file:
                                total_messages += sum(1 for line in file if line.strip())
                        except Exception:
                            pass
                            
    return {
        "session_count": session_count,
        "total_messages_stored": total_messages
    }

File: src/storage.mjs
import fs from 'fs';
import path from 'path';
import { config } from './config.mjs';

export function getSessionDir(sessionId) {
  // Sanitize session ID to prevent path traversal
  const safeId = sessionId.replace(/[^a-zA-Z0-9_\-]/g, '_');
  const sessionDir = path.join(config.data_dir, safeId);
  fs.mkdirSync(sessionDir, { recursive: true });
  return sessionDir;
}

export function getActiveFile(sessionDir) {
  return path.join(sessionDir, "active.jsonl");
}

function _rollFile(sessionDir, activeFile) {
  const existingRolls = [];
  let files;
  try {
    files = fs.readdirSync(sessionDir);
  } catch (e) {
    return; // Can't list dir, skip rolling
  }
  
  for (const file of files) {
    if (file.startsWith("active.") && file.endsWith(".jsonl") && file !== "active.jsonl") {
      const parts = file.split(".");
      if (parts.length === 3) {
        const num = parseInt(parts[1], 10);
        if (!isNaN(num)) existingRolls.push(num);
      }
    }
  }
  const nextNum = (existingRolls.length > 0 ? Math.max(...existingRolls) : 0) + 1;
  const rolledName = `active.${nextNum.toString().padStart(3, '0')}.jsonl`;
  try {
    fs.renameSync(activeFile, path.join(sessionDir, rolledName));
  } catch (e) {
    console.error("Error rolling file:", e.message);
  }
}

export function appendMessages(sessionId, messages) {
  if (!messages || messages.length === 0) return;
  const sessionDir = getSessionDir(sessionId);
  const activeFile = getActiveFile(sessionDir);
  
  try {
    if (fs.existsSync(activeFile)) {
      const stats = fs.statSync(activeFile);
      if (stats.size >= config.roll_size_bytes) {
        _rollFile(sessionDir, activeFile);
      }
    }
  } catch (e) {
    // If stat fails, just keep writing to activeFile
  }
  
  const lines = messages.map(msg => JSON.stringify({ type: "message", message: msg }) + "\n").join("");
  fs.appendFileSync(activeFile, lines, "utf-8");
}

export function getSessionFiles(sessionId) {
  const sessionDir = getSessionDir(sessionId);
  const activeFile = getActiveFile(sessionDir);
  const rolledFiles = [];
  
  try {
    if (fs.existsSync(sessionDir)) {
      const files = fs.readdirSync(sessionDir);
      for (const file of files) {
        if (file.startsWith("active.") && file.endsWith(".jsonl") && file !== "active.jsonl") {
          const parts = file.split(".");
          if (parts.length === 3) {
            const num = parseInt(parts[1], 10);
            if (!isNaN(num)) {
              rolledFiles.push({ num, path: path.join(sessionDir, file) });
            }
          }
        }
      }
    }
  } catch (e) {
    // If we can't read the dir, return what we can
  }
  
  rolledFiles.sort((a, b) => a.num - b.num);
  const allFiles = rolledFiles.map(r => r.path);
  
  try {
    if (fs.existsSync(activeFile)) {
      allFiles.push(activeFile);
    }
  } catch (e) {
    // Skip if can't check
  }
  
  return allFiles;
}

export function getStats() {
  let sessionCount = 0;
  let totalMessages = 0;
  
  try {
    if (fs.existsSync(config.data_dir)) {
      const sessions = fs.readdirSync(config.data_dir);
      for (const session of sessions) {
        const sessionDir = path.join(config.data_dir, session);
        try {
          if (!fs.statSync(sessionDir).isDirectory()) continue;
        } catch (e) {
          continue;
        }
        sessionCount++;
        try {
          const files = fs.readdirSync(sessionDir);
          for (const f of files) {
            if (f.endsWith(".jsonl")) {
              try {
                const content = fs.readFileSync(path.join(sessionDir, f), "utf-8");
                const lines = content.split("\n");
                totalMessages += lines.filter(l => l.trim().length > 0).length;
              } catch (e) {
                // Skip unreadable files
              }
            }
          }
        } catch (e) {
          // Skip unreadable session dir
        }
      }
    }
  } catch (e) {
    // data_dir doesn't exist or not accessible
  }
  
  return { session_count: sessionCount, total_messages_stored: totalMessages };
}
